### DISCUSSION TITLE
Transaction commit behaviour of direct client

### DISCUSSION THREAD

**Discussion by sharangthorat** [SAP/node-rfc#286] [Q&A] [answered]
I am using node_rfc version 2.7.0 and using direct client. My observation is that when I perform BAPI_MATERIAL_SAVEREPLICA operation I can see material available in SAP server without doing explicit BAPI_TRANSACTION_COMMIT on client.  **As per my understanding node_rfc by default is statefull client. However it needs explicit BAPI_TRANSACTION_COMMIT on same connection for data to get committed in SAP**. 

Below is my sample client code. As I have explained above after executing this code I can search material STHORATMATERIAL008 in my SAP server. 

Is my understanding about commit behaviour is correct?  If so why code below works without explicit commit?

```
const Client = require('node-rfc').Client

let client = new Client({
    user: "xxxx",
    passwd: "xxxxx",
    ashost: "xx.xxx.xxx.xx",
    sysnr: "01",
    client: "800",
    lang: "EN"
})


let options = {
    "CLIENTDATA": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "BASE_UOM": "EA"
    }],
    "CLIENTDATAX": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "BASE_UOM": "EA"
    }],
    "MATERIALDESCRIPTION": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "LANGU_ISO": "EN",
        "MATL_DESC": "ENGINE ROD MATERIAL"
    }],
    "PLANTDATA": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "PLANT": "1100",
        "MRP_TYPE": "ND",
        "PROC_TYPE": "E"
    }],
    "PLANTDATAX": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "PLANT": "1100",
        "MRP_TYPE": "X",
        "PROC_TYPE": "X"
    }],
    "UNITSOFMEASURE": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "ALT_UNIT": "EA",
        "ALT_UNIT_ISO": "EA",
        "NUMERATOR": "1",
        "DENOMINATR": "1"
    }],
    "UNITSOFMEASUREX": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "ALT_UNIT": "EA",
        "ALT_UNIT_ISO": "EA",
        "NUMERATOR": "X",
        "DENOMINATR": "X"
    }],
    "HEADDATA": [{
        "FUNCTION": "UPD",
        "MATERIAL": "STHORATMATERIAL008",
        "IND_SECTOR": "M",
        "MATL_TYPE": "HALB",
        "BASIC_VIEW": "X",
        "MRP_VIEW": "X"
    }]
}

console.log('Trying to connect client')

client.connect(function (err,res) {

    if (err) {
        console.log('Error connecting to client', err)
    }else{

        console.log('Trying to invoke client')
       
        client.invoke("BAPI_MATERIAL_SAVEREPLICA", options, function (err, res) {

            if (err) {
                console.log('Error invoking client', err)
              }else
              {
                console.log(' material saved ', res) 
              }
            
        })
    }
 
});
```

**Comment by bsrdjan** [comment, marked as answer]
Hello @sharangthorat,

> As per my understanding node_rfc by default is statefull client. However it needs explicit BAPI_TRANSACTION_COMMIT on same connection for data to get committed in SAP.

I am not sure if BAPI_TRANSACTION_COMMIT is indeed required for BAPI_MATERIAL_SAVEREPLICA. It is not mentioned in BAPI_MATERIAL_SAVEREPLICA documentation and in any case node-rfc does not add (or take) anything to its logic. The node-rfc behaves exactly as ABAP client and if explicit COMMIT not needed in ABAP, the same is with node-rfc. 

I would suggest to post the question in SAP Community, if explicit COMMIT BAPI is required for BAPI_MATERIAL_SAVEREPLICA, how to do rollback etc, see https://answers.sap.com/questions/9220169/doubt-about-bapimaterialsavereplica-and-commit--ro.html

There are also TESTRUN and NO_ROLLBACK_WORK import parameters, influencing the transaction behaviour of this BAPI but better to check with SAP Comunity ABAP experts.
